flashinfer/mla.py

```diff
@@ -0,0 +1,141 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import functools
+from types import SimpleNamespace
+from typing import List, Optional, Tuple, Union
+
+import torch
+
+from .jit import gen_batch_mla_module, get_batch_mla_uri
+from .utils import MaskMode, get_cuda_stream, register_custom_op, register_fake_op
+
+_batch_mla_modules = {}
+
+
+def get_batch_mla_module(*args):
+    global _batch_mla_modules
+    if args not in _batch_mla_modules:
+        _batch_mla_modules[args] = gen_batch_mla_module(*args)
+    return _batch_mla_modules[args]
+
+
+class BatchMLAPageAttentionWrapper:
+    def __init__(
+        self,
+        float_workspace_buffer: torch.Tensor,
+        backend: str = "fa2",
+    ) -> None:
+        r"""Constructor for BatchMLAPageAttentionWrapper."""
+        self._float_workspace_buffer = float_workspace_buffer
+        self.device = float_workspace_buffer.device
+        self._int_workspace_buffer = torch.empty(
+            (8 * 1024 * 1024,), dtype=torch.uint8, device=self.device
+        )
+        self._pin_memory_int_workspace_buffer = torch.empty(
+            self._int_workspace_buffer.shape,
+            dtype=self._int_workspace_buffer.dtype,
+            pin_memory=True,
+        )
+
+    def plan(
+        self,
+        qo_indptr: torch.Tensor,
+        kv_indptr: torch.Tensor,
+        kv_indices: torch.Tensor,
+        kv_len_arr: torch.Tensor,
+        num_heads: torch.Tensor,
+        head_dim_ckv: int,
+        head_dim_kpe: int,
+        page_size: int,
+        causal: bool,
+        sm_scale: float,
+        q_data_type: torch.dtype,
+        kv_data_type: torch.dtype,
+    ) -> None:
+        self._cached_module = get_batch_mla_module(
+            q_data_type,
+            kv_data_type,
+            q_data_type,
+            qo_indptr.dtype,
+            head_dim_ckv,
+            head_dim_kpe,
+        )
+        qo_indptr_host = qo_indptr.to("cpu")
+        kv_indptr_host = kv_indptr.to("cpu")
+        kv_len_arr_host = kv_len_arr.to("cpu")
+
+        self._qo_indptr_buf = qo_indptr
+        self._kv_indptr_buf = kv_indptr
+        self._kv_indices_buf = kv_indices
+        self._kv_len_arr_buf = kv_len_arr
+        self._causal = causal
+        self._page_size = page_size
+        self._sm_scale = sm_scale
+
+        with self.device as device:
+            self._plan_info = self._cached_module.plan(
+                self._float_workspace_buffer,
+                self._int_workspace_buffer,
+                self._pin_memory_int_workspace_buffer,
+                qo_indptr_host,
+                kv_indptr_host,
+                kv_len_arr_host,
+                num_heads,
+                head_dim_ckv,  # head_dim_o
+                causal,
+                get_cuda_stream(device),
+            )
+
```

Comment:
Add these two overloads to make mypy happy.

```suggestion

    @overload
    def run(
        self,
        q_nope: torch.Tensor,
        q_pe: torch.Tensor,
        ckv_cache: torch.Tensor,
        kpe_cache: torch.Tensor,
        return_lse: Literal[False] = False,
    ) -> torch.Tensor:
        ...
    
    @overload
    def run(
        self,
        q_nope: torch.Tensor,
        q_pe: torch.Tensor,
        ckv_cache: torch.Tensor,
        kpe_cache: torch.Tensor,
        return_lse: Literal[True] = True,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        ...

```